[X] don't expand types to Extension for x:Static

StephaneDelcroix opened the pull request:
### Description of Change

if a type name exists with an Extension suffix, xaml rules require us to use that. But not for x:Static


### Issues Fixed

- fixes #11833



<!-- Enter description of the fix in this section -->


<!-- Please make sure that there is a bug logged for the issue being fixed. The bug should describe the problem and how to reproduce it. -->


<!--
Are you targeting main? All PRs should target the main branch unless otherwise noted.
-->

samhouts commented:
Do we know if this is a path that is commonly used? Is this going to break a bunch of people?

StephaneDelcroix commented:
> Do we know if this is a path that is commonly used? Is this going to break a bunch of people?

it's only affecting namespace with both MyClass and MyClassExtension defined. I believe this is safe and the modified behaviour is the expected one

StephaneDelcroix commented:
this is a dupe pr. the dupe is #15903. pick a single one

mattleibow commented:
@StephaneDelcroix this has been many coffee cycles since this was created... Do we still want/need this?

StephaneDelcroix commented:
> @StephaneDelcroix this has been many coffee cycles since this was created... Do we still want/need this?

y, but in net9

mattleibow commented on src/Controls/tests/Xaml.UnitTests/LoaderTests.cs:
```diff
@@ -792,7 +792,7 @@ public void StyleWithoutTargetTypeThrows()
 		// https://github.com/xamarin/Microsoft.Maui.Controls/issues/3606#issuecomment-422377338
 		{
-			var bindingType = XamlParser.GetElementType(new XmlType("http://schemas.microsoft.com/dotnet/2021/maui", "Binding", null), null, null, out var ex);
+			var bindingType = XamlParser.GetElementType(new XmlType("http://schemas.microsoft.com/dotnet/2021/maui", "Binding", null), null, null, true, out var ex);
```
Is this single line change sufficient tests for this PR? Also, that link goes nowhere and seems to make some sort of indication that this test is specific.

Is there a need to check cases where a Type and TypeExtension both exist are OK?

StephaneDelcroix replied:
Binding and BindingExtension both exists

mattleibow reviewed (changes requested):
This PR has ABI breaks as you noted, however, even if we switch to a new TFM (net10) old libraries that use this type and are built with an older maui will still break. People would have to recompile.

One workaround we can do is use default interface members.

Unless... and you would be on the hook for this (not sure what the hook would be and/or if it is enforceable) you say that this type is not meant to be used by anyone outside of maui and doing so is them opting into breaks.

And then this needs that oh-so-common rebase.

StephaneDelcroix commented:
I reviewed this again. the only risk of breaking change would be if a 3rd party provide a markup extension that invoke `serviceproviderGetService(typeof(IXamlTypeResolver))` which is unlikely

StephaneDelcroix committed:
[X] Don't expand types to *Extension for x:Static

if a type name exists with an Extension suffix, xaml rules requires us
to use that, but for x:Static we expect the user to know better

StephaneDelcroix commented:
no longer a breaking change

StephaneDelcroix commented:
/azp run

StephaneDelcroix commented:
/azp run

PureWeen commented:
/rebase
